src/ir_pattern.py

```diff
@@ -0,0 +1,423 @@
+    """
+    Template for defining "IR" patterns that can match against input asm.
+    The matching process uses the FlowGraph and register analysis to compute
+    inter-instruction dependencies, so these patterns can match even when
+    they have been interleaved/reordered by the compiler in the input asm.
+
+    IrPattern subclasses *must* define `parts` and `replacement`, and can
+    optionally implement `check()`.
+
+    For now, the pattern cannot contain any branches, and the replacement
+    must be a single instruction (though, it can be fictive).
+    """
+
+    parts: ClassVar[List[str]]
+    replacement: ClassVar[str]
+
+    def check(self, m: "IrMatch") -> bool:
+        """Override to perform additional checks/calculations before replacement."""
+        return True
+
+    def compile(self, arch: ArchFlowGraph) -> "CompiledIrPattern":
+        missing_meta = InstructionMeta.missing()
+        regf = RegFormatter()
+        replacement_instr = parse_instruction(
+            self.replacement, missing_meta, arch, regf
+        )
+
+        name = f"__pattern_{self.__class__.__name__}"
+        func = Function(name=name)
+        # Add a fictive nop instruction for each input to the replacement_instr
+        # This acts as a placeholder Reference to represent where the input was set
+        assert len(replacement_instr.inputs) == len(
+            set(replacement_instr.inputs)
+        ), "pattern inputs must be unique"
+        for inp in replacement_instr.inputs:
+            func.new_instruction(
+                Instruction(
+                    "in.fictive",
+                    [],
+                    meta=missing_meta,
+                    inputs=[],
+                    clobbers=[],
+                    outputs=[inp],
+                    reads_memory=False,
+                    writes_memory=False,
+                )
+            )
+        for part in self.parts:
+            func.new_instruction(parse_instruction(part, missing_meta, arch, regf))
+
+        asm_data = AsmData()
+        flow_graph = build_flowgraph(func, asm_data, arch, fragment=True)
+        return CompiledIrPattern(
+            source=self,
+            flow_graph=flow_graph,
+            replacement_instr=replacement_instr,
+        )
+
+
+@dataclass(eq=False, frozen=True)
+class CompiledIrPattern:
+    source: IrPattern
+
+    flow_graph: FlowGraph
+    replacement_instr: Instruction
+
+
+@dataclass
+class IrMatch:
+    """
+    IrMatch represents the matched state of an IrPattern.
+    This object is considered read-only; none of its methods modify its state.
+    Its `map_*` methods take a pattern part and return the matched instruction part.
+    """
+
+    symbolic_registers: Dict[str, Register] = field(default_factory=dict)
+    symbolic_labels: Dict[str, str] = field(default_factory=dict)
+    symbolic_args: Dict[str, Argument] = field(default_factory=dict)
+    ref_map: Dict[Reference, RefSet] = field(default_factory=dict)
```

Comment:
"prologue" can be used as a key if you were to reference a register directly in the pattern (e.g. use `$r2` instead of `$k`)

Reply:
Ah! I think we want prologue to always map to prologue, then?

Reply:
Not always - in the pattern, you could use a register like `$r1` or `$cr0` which is not set as part of the pattern, but *is* set by some instruction in the candidate function. 
This happens with `$r1` in the existing patterns: for the pattern, it's source will be the prologue; but in the candidate function it's source will be an `addi r1, r1, N` instruction. 

I had to modify `PrologueRef` & `EpilogueRef` to create unique instances for each Location. (I have been going back-and-forth on whether or not this is necessary, but it is needed in this case.)

I'm also using "PrologueRef" to mean "anything *before* the function," but I'm worried this might get confused with the concept of a *function prologue* where registers are saved -- is this too confusing? I could pick a better pair of terms

Reply:
They're not fully interchangeable, but they are close. There is a PrologueRef for every register in `arch.all_regs`, but there isn't a PrologueRef for symbolic registers like `$a`. 

(I admit this *does* tie back to your other point about uncoupling the pattern `parts` from `replacement_instr`'s inputs, and if the "regs were read before being written to" is actually necessary -- but as-is they enable that current behavior)

Reply:
Mmm. It does bother me that that very minor distinction is enough to create two separate ways of dealing with the same thing. There's one additional difference where with in.fictive's we make temporary registers with copies of the inputs, while with static registers we don't. (And it would be a bit tricky to do so: rename_reg doesn't work for non-symbolic regs, and heuristics in translate.py that depend on specific static regs ($sp in particular) would break. We could add an additional guard check that static regs used in the output instruction still have the same sources, I guess... in practice this won't come up though, ofc. Unless we find some pattern that only works with specific regs but the regs do change across the function.

One ambitious thing we could (later) do in this direction is add a pre-pass that removes stack pointer changes from the asm. Then all static regs would become *really* static and we could go back to a PrologueRef without location argument.

Reply:
Yeah, when I was experimenting I tried creating a temp for the statics as well: making a temp for `$sp` ended up confusing translate.py and affected the output. (Though, I didn't investigate... it might have been an easy fix.)